### PR TITLE
Fix random seed

### PR DESCRIPTION
**Pull request by vaaaaanquish**
Added automatic setting of seed value according to hash, and dump of seed value.
Available in competitions like Kaggle :)

issue: https://github.com/m3dev/gokart/issues/116

### CODE REVIEW

File: gokart/task.py [hunk 1]
```diff
@@ -279,6 +281,34 @@ def get_task_params(self) -> Dict:
...
+        }
+        random.seed(random_seed)
+        np.random.seed(random_seed)
+        try:
+            import torch
+            torch.manual_seed(random_seed)
+            seeds['seed_method'].append('torch.manual_seed')
+            torch.cuda.manual_seed(random_seed)
+            seeds['seed_method'].append('torch.cuda.manual_seed')
+        except Exception:
```
Comment (nishiba) [inline gokart/task.py]: I think ImportError should be used. How do you think?

Comment (nishiba) [reply gokart/task.py]: How do I add other modules in my local project?  Do I need to add codes to gokart?

Comment (vaaaaanquish) [reply gokart/task.py]: Good comment.  
I found two issues in response to your point.

1. add other modules in local project (nishiba-san say)
2. misleading if we want to fix the seed from outside

Change to WIP once.

**Comment by vaaaaanquish** [comment]
@nishiba @Hi-king Please check PR.

I add `fix_random_seed_methods` parameter.
This parameter will enter `fix seed method` list.
For example...
```
fix_random_seed_methods=["random.seed", "torch.random.manual_seed"]
```
If have not installed module in fix_random_seed_methods, that will be `pass`.

File: gokart/task.py (hunk 1 above)
Comment (vaaaaanquish) [reply gokart/task.py]: I add `fix_random_seed_methods` :)

File: gokart/task.py [hunk 2]
```diff
@@ -279,6 +280,36 @@ def get_task_params(self) -> Dict:
...
+                m(random_seed)
+                success_methods.append(method_name)
+            except ModuleNotFoundError:
+                pass
+            except AttributeError:
+                pass
+        return success_methods
+
+    def _get_random_seed(self):
+        return int(self.make_unique_id(), 16) % (2**32 - 1)  # maximum numpy.random.seed
```
Comment (Hi-king) [inline gokart/task.py]: 🆒

File: examples/sample_fix_random_seed.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    #     0.6191630959510803, -0.13227537274360657, -0.6677181720733643]}
+    #
+    # //--------------------- Not install pytorch ----------------------
+    # {'random': [60, 27, 7, 56, 91, 38, 72, 37, 69, 51],
+    #  'numpy': [20, 17, 22, 65, 87, 4, 30, 76, 81, 31],
+    #  'torch': []}
+    # //----------------------------------------------------------------
+    gokart.run([
+        'sample_fix_random_seed.SampleTask', '--local-scheduler', '--rerun', '--sample-param=a',
+        '--fix-random-seed-methods=["random.seed","numpy.random.seed","torch.random.manual_seed"]'
```
Comment (Hi-king) [inline examples/sample_fix_random_seed.py]: How about setting these methods as default random seed methods?

File: gokart/task.py [hunk 4]
```diff
@@ -279,6 +280,36 @@ def get_task_params(self) -> Dict:
             return self.load(target)
         return dict()
 
+    @luigi.Task.event_handler(luigi.Event.START)
+    def _set_random_seed(self):
+        random_seed = self._get_random_seed()
+        seed_methods = self.try_set_seed(self.fix_random_seed_methods, random_seed)
+        self.dump({'seed': random_seed, 'seed_methods': seed_methods}, self._get_random_seeds_target())
```
Comment (Hi-king) [inline gokart/task.py]: Cool way to reproduce results.

IMO, there must be another way to inject random seeds because  task id occasionally changes.
How about introducing another params like `fix_random_seed_value`?

File: examples/sample_fix_random_seed.py (hunk 3 above)
Comment (Hi-king) [reply examples/sample_fix_random_seed.py]: Sorry, I overlooked `default=['random.seed', 'numpy.random.seed']`. Cool!

File: gokart/task.py (hunk 4 above)
Comment (vaaaaanquish) [reply gokart/task.py]: That's right. It looks like a `fixed task_id`. I'll try :)

Comment (vaaaaanquish) [reply gokart/task.py]: done https://github.com/m3dev/gokart/pull/128/commits/068c50819ffa0b401e1be8451b84d3f71715fc64

**Comment by Teppei-Kanayama** [comment]
@vaaaaanquish LGTM!

File: gokart/task.py (hunk 4 above)
Comment (vaaaaanquish) [reply gokart/task.py]: @Hi-king please check PR!

Comment (Hi-king) [reply gokart/task.py]: @vaaaaanquish cool way!
